### PR TITLE
Add support for prefix matching routing

### PR DESCRIPTION
**Pull request by resios**
This allows customers that have inline Swagger definitions with proxy setup to work with SAM local.

For example if a customer has a lambda mounted at `/foo` with an inline Swagger that has an API `/bar` previously requests to `/foo/bar` would have resulted in a `404` response, but now the lambda is invoked as expected.

To use this functionality, the customer must opt-in on it using `--prefix-routing` flag

This also adds support for:
* correctly handling multi-line output (addresses #45)
* correctly handling non-existent static directories
* correctly marshaling event fields that are empty

### CODE REVIEW

File: router/function_test.go [hunk 1]
```diff
@@ -133,9 +133,78 @@ var _ = Describe("Function", func() {
...
+		It("should respond with a 404 to HTTP requests on an invalid path", func() {
+			req, _ := http.NewRequest("GET", "/invalid", nil)
+			rr := httptest.NewRecorder()
+			r.Router().ServeHTTP(rr, req)
+			Expect(rr.Code).To(Equal(http.StatusNotFound))
+		})
+
+		methods := []string{"OPTIONS", "GET", "HEAD", "POST", "PUT", "DELETE", "TRACE", "CONNECT"}
+		for _, method := range methods {
+			It("should respond to HTTP requests on "+method+" /any", func() {
```
Comment (sanathkr) [inline router/function_test.go]: Amazing test! I get super excited when seeing such thorough tests. Love it!

File: router/function_test.go [hunk 2]
```diff
@@ -133,9 +133,78 @@ var _ = Describe("Function", func() {
...
+		r := router.NewServerlessRouter(true)
+
+		function := &cloudformation.AWSServerlessFunction{
+			Runtime: "nodejs6.10",
+			Events: map[string]cloudformation.AWSServerlessFunction_EventSource{
+				"AnyRequests": cloudformation.AWSServerlessFunction_EventSource{
+					Type: "Api",
+					Properties: &cloudformation.AWSServerlessFunction_S3EventOrSNSEventOrKinesisEventOrDynamoDBEventOrApiEventOrScheduleEventOrCloudWatchEventEventOrIoTRuleEventOrAlexaSkillEvent{
+						ApiEvent: &cloudformation.AWSServerlessFunction_ApiEvent{
+							Path:   "/any",
```
Comment (sanathkr) [inline router/function_test.go]: Actually, prefix routing should be automatically enabled when route is `/foo/bar/{path}+`. See: https://aws.amazon.com/blogs/aws/api-gateway-update-new-features-simplify-api-development/. 

It should be fairly simple change to make: In `router.go`, just check for path containing catch-alls and enable prefix routing on the prefix ie. `/foo/bar/{path}+` would enable prefix routing on `/foo/bar`

File: runtime.go [hunk 3]
```diff
@@ -575,6 +532,66 @@ func (r *Runtime) InvokeHTTP(profile string) func(http.ResponseWriter, *http.Req
 
 }
 
+// parseOutput decodes the proxy response from the output of the function and returns
+// the rest
+func parseOutput(w http.ResponseWriter, stdoutTxt io.Reader, runtime string, wg *sync.WaitGroup) (output []byte) {
```
Comment (sanathkr) [inline runtime.go]: Thank you!!

File: runtime_test.go [hunk 4]
```diff
@@ -46,5 +50,111 @@ var _ = Describe("sam", func() {
 
 		})
 
+		Context("parse output", func() {
+			var wg sync.WaitGroup
+			var out []byte
+			var r *fakeResponse
+
+			// cases: error on output, non-json, empty output, missing bodym headers or status,
```
Comment (sanathkr) [inline runtime_test.go]: ❤️

**Comment by sanathkr** [comment]
I think this PR fixes more stuff: #65, #39, #141.

File: router/function_test.go (hunk 2 above)
Comment (resios) [reply router/function_test.go]: Updated pull request to include this functionality. Basically if we detect that a path ends with a catch all variables (e.g. `{foo+}` we instruct the mount to use prefix matching.

I added this functionality to the ServerlessRouterMount to better decouple the implementation (i.e. router) from the metadata (i.e. mount).

Let me know what you think.

**Pull request merged**